plugins/bcm57xx/fu-bcm57xx-device.c

```diff
@@ -0,0 +1,860 @@
+	drvinfo.cmd = ETHTOOL_GDRVINFO;
+	strncpy (ifr.ifr_name, self->ethtool_iface, IFNAMSIZ - 1);
+	ifr.ifr_data = (char *) &drvinfo;
+#ifdef HAVE_IOCTL_H
+	rc = ioctl (self->ethtool_fd, SIOCETHTOOL, &ifr);
+#else
+	g_set_error (error,
+		     FWUPD_ERROR,
+		     FWUPD_ERROR_NOT_SUPPORTED,
+		     "Not supported as <sys/ioctl.h> not found");
+	return FALSE;
+#endif
+	if (rc < 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "cannot get driver information [%i]", rc);
+		return FALSE;
+	}
+	g_debug ("FW version %s", drvinfo.fw_version);
+
+	/* sanity check */
+	if (drvinfo.eedump_len != fu_device_get_firmware_size_max (FU_DEVICE (self))) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "EEPROM size invalid, got 0x%x, expected 0x%x",
+			     drvinfo.eedump_len,
+			     (guint) fu_device_get_firmware_size_max (FU_DEVICE (self)));
+		return FALSE;
+	}
+	if (address + bufsz_wrds * sizeof(guint32) > drvinfo.eedump_len) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "tried to read outside of EEPROM size [0x%x]",
+			     drvinfo.eedump_len);
+		return FALSE;
+	}
+
+	/* read EEPROM (NVRAM) data */
+	eepromsz = sizeof(struct ethtool_eeprom) + bufsz_wrds * sizeof(guint32);
+	eeprom = (struct ethtool_eeprom *) g_malloc0 (eepromsz);
+	eeprom->cmd = ETHTOOL_GEEPROM;
+	eeprom->len = bufsz_wrds * sizeof(guint32);
+	eeprom->offset = address;
+	ifr.ifr_data = (char *) eeprom;
+#ifdef HAVE_IOCTL_H
+	rc = ioctl (self->ethtool_fd, SIOCETHTOOL, &ifr);
+#else
+	g_set_error (error,
+		     FWUPD_ERROR,
+		     FWUPD_ERROR_NOT_SUPPORTED,
+		     "Not supported as <sys/ioctl.h> not found");
+	return FALSE;
+#endif
+	if (rc < 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "cannot read eeprom [%i]", rc);
+		return FALSE;
+	}
+
+	/* copy back data */
+	if (!fu_memcpy_safe ((guint8 *) buf, bufsz_wrds * sizeof(guint32), 0x0,	/* dst */
+			     (guint8 *) eeprom, eepromsz,			/* src */
+			     G_STRUCT_OFFSET(struct ethtool_eeprom, data),
+			     bufsz_wrds * sizeof(guint32), error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+#else
+	g_set_error (error,
+		     FWUPD_ERROR,
+		     FWUPD_ERROR_NOT_SUPPORTED,
+		     "Not supported as <linux/ethtool.h> not found");
+	return FALSE;
+#endif
+}
+
+static gboolean
+fu_bcm57xx_device_nvram_read (FuBcm57xxDevice *self,
+			      guint32 address, guint32 *buf, gsize bufsz,
+			      GError **error)
+{
+	/* simpler! */
+	if (self->ethtool_iface != NULL) {
+		return fu_bcm57xx_device_nvram_read_ethtool (self, address,
+							     buf, bufsz,
+							     error);
+	}
+
+	for (guint i = 0; i < bufsz; i++) {
+		RegNVMCommand_t tmp = { 0 };
+		fu_bcm57xx_device_nvram_clear_done (self);
+		fu_bcm57xx_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+					     REG_NVM_ADDR, address);
+		tmp.bits.Doit = 1;
+		tmp.bits.First = (i == 0);
+		tmp.bits.Last = (i == bufsz - 1);
```

Comment:
I had problems when I originally wrote the code in the fw repo with this. Changing this to write one word at a time fixes the data corruption issue I was seeing before, although it's not the fastest.

I'll see if I can get burst writes working, but for now changing this to the following gets writes working:
tmp.bits.First = 1;
tmp.bits.Last = 1;

Reply:
Note that I'll also work on updating my repo to use ethtool for writing, maybe switching to that here makes sense too.

Reply:
Many thanks, changed.

Reply:
> although it's not the fastest

It took over 16 minutes to write the image using single word writes; I'd hate to think what's this is doing to the flash write limit if it's doing something like a 1024 block read, erase, write for every dword..

Reply:
I just switched the default to use the ethtool API for writing. I think that's a lot safer if there's a data corruption issue using raw writes.